src/py21cmfast/src/PerturbHaloField.c

```diff
@@ -0,0 +1,327 @@
+
+            printf("(new): x = %f y = %f z = %f mass = %e\n",xf,yf,zf,halo_mass[i_halo]);
+
+            xf = xf * user_params->HII_DIM;
+            yf = yf * user_params->HII_DIM;
+            zf = zf * user_params->HII_DIM;
+
+            halos_perturbed->halos_perturbed[HII_R_INDEX(xf,yf,zf)] = halo_mass[i_halo];
+        }
+    }
+
+//    printf("Perturbed halo list\n");
+//    for (i=0; i<user_params->HII_DIM; i++){
+//        for (j=0; j<user_params->HII_DIM; j++){
+//            for (k=0; k<user_params->HII_DIM; k++){
+//                if(halos_perturbed->halos_perturbed[HII_R_INDEX(i,j,k)] > 0.) {
+//                    printf("(new): x = %d y = %d z = %d mass = %e\n",i,j,k,halos_perturbed->halos_perturbed[HII_R_INDEX(i,j,k)]);
+//                }
+//            }
+//        }
+//    }
+
+    // Divide out multiplicative factor to return to pristine state
+#pragma omp parallel shared(boxes,growth_factor,dimension,displacement_factor_2LPT) private(i,j,k) num_threads(user_params->N_THREADS)
+    {
+#pragma omp for
+        for (i=0; i<dimension; i++){
+            for (j=0; j<dimension; j++){
+                for (k=0; k<dimension; k++){
+//                    if(user_params->PERTURB_ON_HIGH_RES) {
+//                        boxes->hires_vx[R_INDEX(i,j,k)] /= growth_factor / user_params->BOX_LEN;
+//                        boxes->hires_vy[R_INDEX(i,j,k)] /= growth_factor / user_params->BOX_LEN;
+//                        boxes->hires_vz[R_INDEX(i,j,k)] /= growth_factor / user_params->BOX_LEN;
+//                        if(global_params.SECOND_ORDER_LPT_CORRECTIONS){
+//                            boxes->hires_vx_2LPT[R_INDEX(i,j,k)] /= displacement_factor_2LPT / user_params->BOX_LEN;
+//                            boxes->hires_vy_2LPT[R_INDEX(i,j,k)] /= displacement_factor_2LPT / user_params->BOX_LEN;
+//                            boxes->hires_vz_2LPT[R_INDEX(i,j,k)] /= displacement_factor_2LPT / user_params->BOX_LEN;
+//
+//                        }
+//                    }
+//                    else {
+//                        boxes->lowres_vx[HII_R_INDEX(i,j,k)] /= growth_factor / user_params->BOX_LEN;
+//                        boxes->lowres_vy[HII_R_INDEX(i,j,k)] /= growth_factor / user_params->BOX_LEN;
+//                        boxes->lowres_vz[HII_R_INDEX(i,j,k)] /= growth_factor / user_params->BOX_LEN;
+//                        if(global_params.SECOND_ORDER_LPT_CORRECTIONS){
+//                            boxes->lowres_vx_2LPT[HII_R_INDEX(i,j,k)] /= displacement_factor_2LPT / user_params->BOX_LEN;
+//                            boxes->lowres_vy_2LPT[HII_R_INDEX(i,j,k)] /= displacement_factor_2LPT / user_params->BOX_LEN;
+//                            boxes->lowres_vz_2LPT[HII_R_INDEX(i,j,k)] /= displacement_factor_2LPT / user_params->BOX_LEN;
+//
+//                        }
+//                    }
+                    boxes->lowres_vx[HII_R_INDEX(i,j,k)] /= growth_factor / user_params->BOX_LEN;
```

Comment:
could pre-calculate `growth_factor/BOX_LEN` here to save some cycles. Also, calculate `HII_R_INDEX` once each iter.